nixos.nix

```diff
@@ -4,454 +4,184 @@ let
-                  # Save the default permissions at the level the
-                  # directory resides. This used when creating its
-                  # parent directories, giving them reasonable
-                  # default permissions unaffected by the
-                  # directory's own.
-                  defaultPerms = mapAttrs (_: x: x // { internal = true; }) dirPermsOpts;
-                  dirPath = mkOption {
-                    type = path;
-                    internal = true;
-                  };
-                } // dirPermsOpts;
-              };
-              rootFile = submodule [
-                commonOpts
-                fileOpts
-                ({ config, ... }: {
-                  parentDirectory = mkDefault (defaultPerms // rec {
-                    directory = dirOf config.file;
-                    dirPath = directory;
-                    inherit (config) persistentStoragePath;
-                    inherit defaultPerms;
-                  });
-                  filePath = mkDefault config.file;
-                })
-              ];
-              rootDir = submodule ([
-                commonOpts
-                dirOpts
-                ({ config, ... }: {
-                  defaultPerms = mkDefault defaultPerms;
-                  dirPath = mkDefault config.directory;
-                })
-              ] ++ (mapAttrsToList (n: v: { ${n} = mkDefault v; }) defaultPerms));
-            in
-            {
-              options =
-                {
-                  enable = mkOption {
-                    type = bool;
-                    default = true;
-                    description = "Whether to enable this persistent storage location.";
-                  };
-
-                  persistentStoragePath = mkOption {
-                    type = path;
-                    default = name;
-                    defaultText = "‹name›";
-                    description = ''
-                      The path to persistent storage where the real
-                      files and directories should be stored.
-                    '';
-                  };
-
-                  users = mkOption {
-                    type = attrsOf (
-                      submodule (
-                        { name, config, ... }:
-                        let
-                          userDefaultPerms = {
-                            inherit (defaultPerms) mode;
+                  users =
+                    let
+                      outerName = name;
+                      outerConfig = config;
+                    in
+                    mkOption {
+                      type = attrsOf (
+                        submodule (
+                          { name, config, ... }:
+                          import ./submodule-options.nix {
+                            inherit pkgs lib;
+                            config = outerConfig // config;
+                            name = outerName;
+                            usersOpts = true;
```

Comment:
Again, this feels like it's fighting the module system.

I'm not too keen on `config = outerConfig // config` either, surely it's better to be explicit about those being separate?